app/Models/User.php

```diff
@@ -495,21 +496,57 @@ public static function lookupWithHistory($usernameOrId, $type = null, $findAll =
         }
     }
 
-    public function addToGroup(Group $group, ?self $actor = null): void
+    public function addToGroup(Group $group, ?array $playmodes = null, ?self $actor = null): void
     {
-        if ($this->findUserGroup($group, true) !== null) {
+        $playmodes = array_unique($playmodes ?? []);
+
+        if (!$group->has_playmodes && $playmodes !== []) {
+            throw new InvariantException('Group does not allow playmodes');
+        }
+
+        $invalidPlaymodes = array_diff($playmodes, array_keys(Beatmap::MODES));
+
+        if ($invalidPlaymodes !== []) {
+            throw new InvariantException('Invalid playmodes: '.implode(', ', $invalidPlaymodes));
+        }
+
+        $activeUserGroup = $this->findUserGroup($group, true);
+
+        if (
+            $activeUserGroup !== null &&
+            (new Set($playmodes))->xor(new Set($activeUserGroup->playmodes ?? []))->isEmpty()
+        ) {
             return;
         }
 
-        $this->getConnection()->transaction(function () use ($actor, $group) {
-            $this
-                ->userGroups()
-                ->firstOrNew(['group_id' => $group->getKey()])
-                ->fill(['user_pending' => false])
+        $this->getConnection()->transaction(function () use ($actor, $group, $playmodes, $activeUserGroup) {
+            if ($activeUserGroup === null) {
+                UserGroupEvent::logUserAdd($actor, $this, $group, $playmodes);
+            } else {
+                $previousPlaymodes = $activeUserGroup->playmodes ?? [];
+                $playmodesAdded = array_values(array_diff($playmodes, $previousPlaymodes));
+                $playmodesRemoved = array_values(array_diff($previousPlaymodes, $playmodes));
+
+                if ($playmodesAdded !== []) {
+                    UserGroupEvent::logUserAddPlaymodes($actor, $this, $group, $playmodesAdded);
+                }
+
+                if ($playmodesRemoved !== []) {
+                    UserGroupEvent::logUserRemovePlaymodes($actor, $this, $group, $playmodesRemoved);
+                }
+            }
+
+            ($activeUserGroup ??
+                $this->userGroups()->firstOrNew(['group_id' => $group->getKey()])
+            )
+                ->fill([
+                    'playmodes' => $playmodes === [] ? null : $playmodes,
```

Comment:
this probably should check if the group has playmodes instead and assign empty array anyway if has playmodes (or do this earlier to also allow to tell whether the group has playmodes or not in the event log instead of having them all contain an array value)